Content.Server/Roles/RoleTimerSystem.cs

```diff
@@ -0,0 +1,266 @@
+    }
+
+    private void OnPlayerAttached(PlayerAttachedEvent ev)
+    {
+        PlayerRolesChanged(ev.Player);
+    }
+
+    private void OnPlayerDetached(PlayerDetachedEvent ev)
+    {
+        // This doesn't fire if the player doesn't leave their body. I guess it's fine?
+        PlayerRolesChanged(ev.Player);
+    }
+
+    public bool IsAllowed(NetUserId id, string role)
+    {
+        if (!ProtoManager.TryIndex<JobPrototype>(role, out var job) ||
+            job.Requirements == null ||
+            !ConfigManager.GetCVar(CCVars.GameRoleTimers)) return true;
+
+        TimeSpan? overall = null;
+        Dictionary<string, TimeSpan>? roles = null;
+
+        foreach (var requirement in job.Requirements)
+        {
+            var reason = RequirementMet(id, job, requirement, overall, roles);
+            if (reason == null) continue;
+            return false;
+        }
+
+        return true;
+    }
+
+    public HashSet<string> GetDisallowedJobs(NetUserId id)
+    {
+        var roles = new HashSet<string>();
+        if (!ConfigManager.GetCVar(CCVars.GameRoleTimers)) return roles;
+
+        TimeSpan? overallPlaytime = null;
+        Dictionary<string, TimeSpan>? rolePlaytimes = null;
+
+        foreach (var job in ProtoManager.EnumeratePrototypes<JobPrototype>())
+        {
+            if (job.Requirements != null)
+            {
+                foreach (var requirement in job.Requirements)
+                {
+                    var reason = RequirementMet(id, job, requirement, overallPlaytime, rolePlaytimes);
+                    if (reason == null) continue;
+                    goto NoRole;
+                }
+            }
+
+            roles.Add(job.ID);
+            NoRole:;
+        }
+
+        return roles;
+    }
+
+    public void SetAllowedJobs(NetUserId id, ref List<string> jobs)
+    {
+        if (!ConfigManager.GetCVar(CCVars.GameRoleTimers)) return;
+
+        TimeSpan? overallPlaytime = null;
+        Dictionary<string, TimeSpan>? rolePlaytimes = null;
+
+        for (var i = 0; i < jobs.Count; i++)
+        {
+            var job = jobs[i];
+
+            if (!ProtoManager.TryIndex<JobPrototype>(job, out var jobber) ||
+                jobber.Requirements == null ||
+                jobber.Requirements.Count == 0) continue;
+
+            foreach (var requirement in jobber.Requirements)
+            {
+                var reason = RequirementMet(id, jobber, requirement, overallPlaytime, rolePlaytimes);
+                if (reason == null) continue;
+
+                jobs.RemoveSwap(i);
+                i--;
+            }
+        }
+    }
+
+    private List<string> GetRoles(IPlayerSession session)
+    {
+        var roles = new List<string>();
+        var contentData = _playerManager.GetPlayerData(session.UserId).ContentData();
+
+        if (contentData?.Mind == null) return roles;
+
+        foreach (var mindRole in contentData.Mind.AllRoles)
+        {
+            roles.Add(mindRole.Name);
+        }
```

Comment:
I dunno how best to fix this, because roles are weird & not all roles necessarily have prototypes, but this should really be using prototype IDs. Antags might get renamed, and in the case of Jobs, the job-role name is actually a localized string (its `Role.Name == Loc.GetString(JobPrototype.Name)`).

Maybe explicitly try to cast the `Role` to a `Job` or `TraitorRole` and try to get their relevant prototypes, and then use the prototype IDs? I guess explicitly tracking ObserverRole playtime is also worthwhile.

Or maybe just give the role class an ID field, and have Job & traitor point to the prototype's ID?

Reply:
> Maybe explicitly try to cast the Role to a Job or TraitorRole and try to get their relevant prototypes, and then use the prototype IDs?

Won't that have the same issue of ID overlap? The real issue is roles are bad to work with and I don't have the energy to try and refactor half of mind code trying to deal with it.

Reply:
I guess you could somehow extract the prototype name (i.e., `AntagPrototype` -> "antag"), and then use <prototype-name/type>.<prototype-ID> in place of `mindRole.Name`. That's pretty general and should have no overlap?

Or I guess you just need to define a role-prototype that exists just for tracking and give Antag & job roles relevant prototypes? I'm really not sure what the best fix is.